Details overloading `==` and `!=` operators on the `ValueObject` type (#27040)
dotnet/docs: docs/architecture/microservices/microservice-ddd-cqrs-patterns/implement-value-objects.md

}
```

> You can see how this value object implementation of `Address` has no identity and therefore, no ID field, neither at the `Address` class definition nor the `ValueObject` class definition.

Reviewer: Acrolinx advises to avoid neither/nor

```suggestion
This value object implementation of `Address` has no identity, and therefore no ID field is defined for it, either in the `Address` class definition or the `ValueObject` class definition.
```

Having no ID field in a class to be used by Entity Framework (EF) was not possible until EF Core 2.0, which greatly helps to implement better value objects with no ID. That is precisely the explanation of the next section.

It could be argued that value objects, being immutable, should be read-only (that is, have get-only properties), and that's indeed true. However, value objects are usually serialized and deserialized to go through message queues, and being read-only stops the deserializer from assigning values, so you just leave them as `private set`, which is read-only enough to be practical.

### Value object comparison semantics

Two instances of the `Address` type can be compared using all the following methods:

```csharp
var one = new Address("1 Microsoft Way", "Redmond", "WA", "US", "98052");
var two = new Address("1 Microsoft Way", "Redmond", "WA", "US", "98052");

Console.WriteLine(EqualityComparer<Address>.Default.Equals(one, two)); // True
Console.WriteLine(object.Equals(one, two)); // True
Console.WriteLine(one.Equals(two)); // True
Console.WriteLine(one == two); // True
```

> When all the values are the same, the comparisons are correctly evaluated as `true`. If you didn't choose to overload the `==` and `!=` operators, then the last comparison of `one == two` would evaluate as `false`. For more information, see [Overload ValueObject equality operators](#equal-op-overload).

Reviewer: True is capitalized in the code comments but lowercase here.
